pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java

```diff
@@ -36,209 +34,302 @@
-  private long _segmentUploadStartTime = -1;
-  private String _crypterName;
-  private Map<String, String> _customMap;
+  private final ZNRecord _znRecord;
+  private Map<String, String> _simpleFields;
 
-  @Deprecated
-  private String _rawTableName;
+  // Cache start/end time because they can be used to sort the metadata
+  private long _startTimeMs = -1;
+  private long _endTimeMs = -1;
 
-  public SegmentZKMetadata() {
+  public SegmentZKMetadata(String segmentName) {
+    _znRecord = new ZNRecord(segmentName);
+    _simpleFields = _znRecord.getSimpleFields();
+    // TODO: Remove this field after releasing 0.9.0
+    _simpleFields.put(Segment.SEGMENT_NAME, segmentName);
   }
 
   public SegmentZKMetadata(ZNRecord znRecord) {
-    _segmentName = znRecord.getSimpleField(Segment.SEGMENT_NAME);
-    _segmentType = znRecord.getEnumField(Segment.SEGMENT_TYPE, SegmentType.class, SegmentType.OFFLINE);
-    _startTime = znRecord.getLongField(Segment.START_TIME, -1);
-    _endTime = znRecord.getLongField(Segment.END_TIME, -1);
-    String timeUnitString = znRecord.getSimpleField(Segment.TIME_UNIT);
-    if (timeUnitString != null && !timeUnitString.equals(NULL)) {
-      _timeUnit = znRecord.getEnumField(Segment.TIME_UNIT, TimeUnit.class, TimeUnit.DAYS);
-    }
-    _indexVersion = znRecord.getSimpleField(Segment.INDEX_VERSION);
-    _totalDocs = znRecord.getLongField(Segment.TOTAL_DOCS, -1);
-    _crc = znRecord.getLongField(Segment.CRC, -1);
-    _creationTime = znRecord.getLongField(Segment.CREATION_TIME, -1);
-    try {
-      String partitionMetadataJson = znRecord.getSimpleField(Segment.PARTITION_METADATA);
-      if (partitionMetadataJson != null) {
-        _partitionMetadata = SegmentPartitionMetadata.fromJsonString(partitionMetadataJson);
-      }
-    } catch (IOException e) {
-      LOGGER.error(
-          "Exception caught while reading partition info from zk metadata for segment '{}', partition info dropped.",
-          _segmentName, e);
-    }
-    _segmentUploadStartTime = znRecord.getLongField(Segment.SEGMENT_UPLOAD_START_TIME, -1);
-    _crypterName = znRecord.getSimpleField(Segment.CRYPTER_NAME);
-    _customMap = znRecord.getMapField(Segment.CUSTOM_MAP);
-
-    // For backward-compatibility
-    setTableName(znRecord.getSimpleField(Segment.TABLE_NAME));
+    _znRecord = znRecord;
+    _simpleFields = znRecord.getSimpleFields();
   }
 
   public String getSegmentName() {
-    return _segmentName;
-  }
-
-  public void setSegmentName(String segmentName) {
-    _segmentName = segmentName;
-  }
-
-  public SegmentType getSegmentType() {
-    return _segmentType;
-  }
-
-  public void setSegmentType(SegmentType segmentType) {
-    _segmentType = segmentType;
+    return _znRecord.getId();
   }
 
   public long getStartTimeMs() {
-    if (_startTime > 0 && _timeUnit != null) {
-      return _timeUnit.toMillis(_startTime);
-    } else {
-      return -1;
+    if (_startTimeMs < 0) {
```

Comment:
use <= ?

Reply:
Added a boolean field to track whether the time is cached